README.md
# HttpDuplex

> Unifies http request and response streams.
>> There be 🐲 here! The API and functionality are being cemented, anything before a 1.0.0 release is subject to change.

[![Npm Version](https://img.shields.io/npm/v/@allegiant/httpduplex.svg)](https://www.npmjs.com/package/@allegiant/httpduplex)
[![Build Status](https://travis-ci.org/allegiant-js/httpduplex.svg?branch=master)](https://travis-ci.org/allegiant-js/httpduplex.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/allegiant-js/httpduplex/badge.svg?branch=master)](https://coveralls.io/github/allegiant-js/httpduplex?branch=master)

## Install

```
npm install @allegiant/httpduplex
```

## Usage

The following example can be found at example/index.js

```js
const fs = require('fs');
const http = require('http');
const httpDuplex = require('@allegiant/httpduplex');

var server = http.createServer(function (req, res) {
    var dup = new httpDuplex(req, res);
    console.log(dup.method + ' ' + dup.url); // eslint-disable-line

    if (dup.url == '/') {
        dup.setHeader('content-type', 'text/plain');
        if (dup.method === 'POST') {
            let size = 0;
            dup.on('data', buf => size += buf.length)
               .on('end', () => dup.end(size + '\n'));
        } else {
            fs.createReadStream('README.md').pipe(dup);
        }
    } else {
        dup.setHeader('content-type', 'text/html');
        dup.statusCode = 404;
        dup.end('Not found');
    }
})
.on('listening', function () {
    console.log('Listening http://localhost:' + server.address().port + '/'); // eslint-disable-line
})
.listen(7000);
```


## Example

Running the following command will start up a simple http server:

```
node example/index.js
```

## Thanks

This began as a compatiblity rewrite of [http-duplex](https://github.com/substack/http-duplex).

## Copyright & License

Copyright &copy; 2017 Allegiant. Distributed under the terms of the MIT License, see [LICENSE](https://github.com/allegiant-js/httpduplex/blob/master/LICENSE)

Availble via [npm](https://www.npmjs.com/package/@allegiant/httpduplex) or [github](https://github.com/allegiant-js/httpduplex).


CHANGELOG.md
# 0.3.7 (12/29/2017)

- added travis-ci and coverage via coveralls, replaced tape with tap

# 0.3.6 (12/28/2017)

- include example, remove npm prefer global, republished

# 0.3.5 (12/27/2017)

- backported slight update for headers, docs and repo split.

# 0.3.4 (11/10/2017)

- updates duplex lib to fix cork, uncork and add some chaining

# 0.1.0 (11/05/2017)

- Replace http-duplex package replacing w/ compatible replacement lib, 100% compatability ends here.
  With successive newer versions of node, some slight changes may take place to keep current.


test/httpduplex.js
var test = require('tap').test;
var httpDuplex = require('../index');
var http = require('http');
var request = require('request');
var fs = require('fs');
var selfSrc = fs.readFileSync(__filename);

Object.prototype.serialize = function() {
    return JSON.stringify(this, null, 4);
};

Object.prototype.filterKeys = function(key) {
    var obj = this;
    Object.keys(this).forEach(function (i) {
        if (i == key)
            delete obj[i];
    });

    return obj;
};

String.prototype.format = function() {
    var args = Array.from(arguments);
    return this.replace(/{(\d+)}/g, function(match, number) {
        return typeof args[number] != 'undefined' ? args[number] : match;
    });
};

String.prototype.streamlineLineEndings = function(ending = "\n") {
    return this.replace(/[\r\n,\r,\n]+/g, ending);
};

String.prototype.streamlineSpace = function() {
    return this.replace(/[\f\t\v ]{2,}/g, ' ');
};

String.prototype.streamline = function(ending = "\n") {
    return this.streamlineSpace().streamlineLineEndings(ending);
};

var server = http.createServer(function (req, res) {
    var dup = new httpDuplex(req, res);
    console.log(dup.method + ' ' + dup.url); // eslint-disable-line

    switch (dup.url) {
        case '/':
            dup.setHeader('content-type', 'text/plain');
            if (dup.method === 'POST') {
                var size = 0;
                dup.on('data', function (buf) { size += buf.length; });
                dup.on('end', function () {
                    dup.end(size + '\n');
                });
            }
            else fs.createReadStream(__filename).pipe(dup);
        break;
        case '/info':
            if (dup.method == 'GET') {
                dup.setHeader('content-type', 'text/plain');
                var output = (
                    "Method: {0}\n" +
                    "Path: {1}\n" +
                    "Status: {2}\n" +
                    "Upgrade: {3}\n" +
                    "Http Version 1: {4}\n" +
                    "Http Version 2: {5}\n" +
                    "Headers: \n{6}\n" +
                    "Trailers: {7}\n" +
                    "Complete: {8}\n" +
                    "Readable: {9}\n" +
                    "Writeable: {10}\n" +
                    "Connection: {11}\n" +
                    "Client: {12}\n" +
                    "Socket: {13}\n"
                ).format (
                    dup.method,
                    dup.url,
                    dup.statusCode,
                    dup.upgrade,
                    dup.httpVersion,
                    "{0}.{1}".format(dup.httpVersionMajor, dup.httpVersionMinor),
                    dup.headers.filterKeys('undefined').serialize(),
                    dup.trailers.serialize(),
                    dup.complete,
                    dup.readable,
                    dup.writeable,
                    dup.connection,
                    dup.client,
                    dup.socket
                );
                dup.end(output.streamline());
            } else {
                dup.statusCode = 400;
                dup.end("Bad Request");
            }
        break;
        default:
            dup.statusCode = 404;
            dup.end("File doesn't exist");
        break;
    }
});

test('httpduplex', (t) => {
    t.plan(3);

    server.listen(0);
    server.on('listening', function () {
        var u = 'http://localhost:' + server.address().port + '/';

        request(u, function (err, res, body) {
            if (err) t.fail(err);
            t.equal(String(body), String(selfSrc));
        });

        var r = request.post(u, function (err, res, body) {
            if (err) t.fail(err);
            t.equal(body, '10\n');
        });
        r.end('beep boop\n');

        request(u + 'info', function (err, res, body) {
            if (err) t.fail(err);
            var check = "Method: GET\n" +
            "Path: /info\n" +
            "Status: 200\n" +
            "Upgrade: false\n" +
            "Http Version 1: 1.1\n" +
            "Http Version 2: 1.1\n" +
            "Headers: \n" +
            "{\n" +
            " \"host\": \"localhost:" + server.address().port + "\",\n" +
            " \"connection\": \"close\"\n" +
            "}\n" +
            "Trailers: {}\n" +
            "Complete: false\n" +
            "Readable: true\n" +
            "Writeable: {10}\n" +
            "Connection: [object Object]\n" +
            "Client: [object Object]\n" +
            "Socket: [object Object]\n";

            t.equal(String(body.streamline()), String(check.streamline()));
        });
    });

    t.on('end', function () {
        server.close();
    });
});


example/index.js
const fs = require('fs');
const http = require('http');
const httpDuplex = require('../index');

var server = http.createServer(function (req, res) {
    var dup = new httpDuplex(req, res);
    console.log(dup.method + ' ' + dup.url); // eslint-disable-line

    if (dup.url == '/') {
        dup.setHeader('content-type', 'text/plain');
        if (dup.method === 'POST') {
            let size = 0;
            dup.on('data', buf => size += buf.length)
               .on('end', () => dup.end(size + '\n'));
        } else {
            fs.createReadStream('README.md').pipe(dup);
        }
    } else {
        dup.setHeader('content-type', 'text/html');
        dup.statusCode = 404;
        dup.end('<h1>Not found</h1>');
    }
})
.on('listening', function () {
    console.log('Listening http://localhost:' + server.address().port + '/'); // eslint-disable-line
})
.listen(7000);


tryitout.js
const { name, description } = require('./package.json');

module.exports = {
  title: name,
  nav: {
    Source: 'https://github.com/allegiant-js/httpduplex',
    Docs: './code/index.html'
  },
  body: `
      <div style="width:80%;position: absolute;left: 50%;top: 50%;-webkit-transform: translate(-50%, -50%);transform: translate(-50%, -50%);">
        <h3 class="text-center" style="font-weight: 100"> ${description} </h3>
        <pre style="background-color:rgb(7, 7, 7);color:rgb(228, 228, 228);white-space: pre;text-align:left;width: auto;display:inline-block;">
const fs = require('fs');
const http = require('http');
const httpDuplex = require('@allegiant/httpduplex');

var server = http.createServer(function (req, res) {
        var dup = new httpDuplex(req, res);
        console.log(dup.method + ' ' + dup.url);

        if (dup.url == '/') {
            dup.setHeader('content-type', 'text/plain');
            if (dup.method === 'POST') {
                let size = 0;
                dup.on('data', buf => size += buf.length)
                   .on('end', () => dup.end(size.toString()));
            } else {
                fs.createReadStream('README.md').pipe(dup);
            }
        } else {
            dup.setHeader('content-type', 'text/html');
            dup.statusCode = 404;
            dup.end('Not found');
        }
}).on('listening', function () {
    console.log('Listening http://localhost:' + server.address().port + '/');
}).listen(7000);
        </pre>
      </div>
    `,
  options: {
    width: '80%'
  },
  footer: `
       <div class="text-black">Made with ☕️ by <a href="https://github.com/allegiant-js">✭ Allegiant</a></div>
    `,
  template: 'landing',
  output: './docs'
};
